net-im/telegram-desktop/telegram-desktop-5.1.7.ebuild

```diff
@@ -0,0 +1,260 @@
+		>=dev-qt/qtnetwork-5.15:5[ssl]
+		>=dev-qt/qtsvg-5.15:5
+		>=dev-qt/qtwidgets-5.15:5[png,X?]
+		kde-frameworks/kcoreaddons:5
+		webkit? (
+			>=dev-qt/qtdeclarative-5.15:5
+			>=dev-qt/qtwayland-5.15:5[compositor(+)]
+		)
+	)
+	qt6? (
+		>=dev-qt/qtbase-6.5:6=[dbus?,gui,network,opengl,wayland?,widgets,X?]
+		>=dev-qt/qtimageformats-6.5:6
+		>=dev-qt/qtsvg-6.5:6
+		webkit? (
+			>=dev-qt/qtdeclarative-6.5:6
+			>=dev-qt/qtwayland-6.5:6[compositor,qml]
+		)
+		qt6-imageformats? (
+			>=dev-qt/qtimageformats-6.5:6=
+			${KIMAGEFORMATS_RDEPEND}
+		)
+	)
+	X? (
+		x11-libs/libxcb:=
+		x11-libs/xcb-util-keysyms
+	)
+"
+RDEPEND="${CDEPEND}
+	webkit? ( || ( net-libs/webkit-gtk:4.1 net-libs/webkit-gtk:6 ) )
+"
+DEPEND="${CDEPEND}
+	>=dev-cpp/cppgir-2.0_p20240315
+	>=dev-cpp/ms-gsl-4
+	dev-cpp/expected-lite
+	dev-cpp/range-v3
+"
+BDEPEND="
+	${PYTHON_DEPS}
+	>=dev-build/cmake-3.16
+	>=dev-cpp/cppgir-2.0_p20240315
+	dev-util/gdbus-codegen
+	virtual/pkgconfig
+	wayland? ( dev-util/wayland-scanner )
+"
+
+PATCHES=(
+	"${FILESDIR}/tdesktop-4.2.4-jemalloc-only-telegram-r1.patch"
+	"${FILESDIR}/tdesktop-4.10.0-system-cppgir.patch"
+	"${FILESDIR}/tdesktop-5.0.1-qt6-no-wayland.patch"
+)
+
+pkg_pretend() {
+	if has ccache ${FEATURES}; then
+		ewarn "ccache does not work with ${PN} out of the box"
+		ewarn "due to usage of precompiled headers"
+		ewarn "check bug https://bugs.gentoo.org/715114 for more info"
+		ewarn
+	fi
+}
+
+src_prepare() {
+	# Bundle kde-frameworks/kimageformats for qt6, since it's impossible to
+	#   build in gentoo right now.
+	if use qt6-imageformats; then
+		sed -e 's/DESKTOP_APP_USE_PACKAGED_LAZY/TRUE/' -i \
+			cmake/external/kimageformats/CMakeLists.txt || die
+		printf '%s\n' \
+			'Q_IMPORT_PLUGIN(QAVIFPlugin)' \
+			'Q_IMPORT_PLUGIN(HEIFPlugin)' \
+			'Q_IMPORT_PLUGIN(QJpegXLPlugin)' \
+			>> cmake/external/qt/qt_static_plugins/qt_static_plugins.cpp || die
+	fi
+	# kde-frameworks/kcoreaddons is bundled when using qt6.
+
+	# Happily fail if libraries aren't found...
+	find -type f \( -name 'CMakeLists.txt' -o -name '*.cmake' \) \
+		\! -path './Telegram/lib_webview/CMakeLists.txt' \
+		\! -path './cmake/external/expected/CMakeLists.txt' \
+		\! -path './cmake/external/kcoreaddons/CMakeLists.txt' \
+		\! -path './cmake/external/qt/package.cmake' \
+		-print0 | xargs -0 sed -i \
+		-e '/pkg_check_modules(/s/[^ ]*)/REQUIRED &/' \
+		-e '/find_package(/s/)/ REQUIRED)/' || die
+	# Make sure to check the excluded files for new
+	# CMAKE_DISABLE_FIND_PACKAGE entries.
+
+	# Control QtDBus dependency from here, to avoid messing with QtGui.
+	# QtGui will use find_package to find QtDbus as well, which
+	# conflicts with the -DCMAKE_DISABLE_FIND_PACKAGE method.
+	if ! use dbus; then
+		sed -e '/find_package(Qt[^ ]* OPTIONAL_COMPONENTS/s/DBus *//' \
+			-i cmake/external/qt/package.cmake || die
+	fi
+
+	if ! use libdispatch; then
+		sed -e '/add_checked_subdirectory(dispatch)/d' \
+			-i cmake/external/CMakeLists.txt || die
+	fi
```

Comment:
can we provide a cmake option similar as jemalloc? also is it possible to upstream this patch? or in the worst case, we have to carry the fix at downstream level forever ..

I will let @mid-kid comment on this..

Reply:
Thanks a lot for looking into this! It's been bugging me for quite a while but I've not actioned upon it due to lack of time.

Anyway, you're right, I personally prefer patches, wherever it makes sense to do so, and I think this is one of those cases. It's not really hacky like the dbus one, and not temporary like qt6-imageformats. And like you said, it can be submitted upstream, I try to nudge nudge wink wink them to the patches directory from time to time.

That said, I think libdispatch is a dangerous one to "just allow" like that, especially since upstream doesn't have a toggle for it. Have you tested it? Do audio/video calls work? If either of those are false please mask the flag to just allow for testing.

Reply:
@Jamim @mid-kid ping, any plan on how we should proceed?

Reply:
Just mask the USE flag for the time being.

Reply:
Sorry for the delay!
I'm going to implement a corresponding cmake option and test the calls this weekend.

Reply:
Hi @dlan17 and @mid-kid,

I've just tested video calls on `telegram-desktop-5.2.0` that was built with `USE="-libdispatch"`. Audio and video works flawlessly, so I believe we can introduce an ability to disable `libdispatch`, but keep it enabled by default.

Reply:
@Jamim great, that's really good news!
@mid-kid can you check/verify this? and give an "approved-by" if it's ok, otherwise I will take a look few days later..

Reply:
I've checked the diffs and everything looks correct to me, and I trust jamim's testing enough to just push it to the tree at this point, and see if we receive any bug reports.

The only thing I'd ask for is a bump to 5.2.2, which is a bugfix release on top of 5.2.0, so shouldn't have regressed anything.

Reply:
I've updated the PR, so now it contains `5.2.2`.

Reply:
Alright, merging now would be good.

Reply:
great, I will handle this PR..
btw, can you guys consider dropping some old versions maybe in next version bump?

Reply:
Sure, I'll drop some old versions next time.

Reply:
Don't drop 5.0.4, it's the stable target